### PR TITLE
Deactivate prescans on suicide

### LINKED ISSUE #36205: Cloner autoprocess clones suicides
When a cloner runs on autoprocess, it will clone scanned players that have suicided.
If I'm not reading this wrong, clone pods should always eject suicides:

`if((mob_occupant.stat == DEAD) || (mob_occupant.suiciding) || mob_occupant.hellbound)  //Autoeject corpses and suiciding dudes.`

> Jayce Metzer is getting too high on life! It looks like he's trying to commit suicide.
Jayce Metzer seizes up and falls limp, his eyes dead and lifeless...
Jayce Metzer has died at Construction Area.
Jayce Metzer's Memories:
Time of death: 12:54:40
Your body is beginning to regenerate in a cloning pod. You will become conscious when it is complete.

### PR DESCRIPTION
**Pull request by george99g**
This will just give all of your records a flag on suicide, preventing them from being used to create a clone.

Created `suicide_act` on `/datum/mind` to log animal suicides and handle the record deactivation without cluttering up the mob verbs.

Screenshot:
screenshot at 2018-03-27 15-48-14

Fixes #36205 
Alternative to and closes #36839

:cl:
fix: Suiciding will now deactivate all your genetics prescans.
/:cl:

### CODE REVIEW

**Comment by MrDoomBringer** [comment]
I feel like it'd be better to create a note or something in the cloning scanner, rather than just deactivate it completely. I feel like theres a few cases where someone will suicide to accomplish something, rather than to be completely removed from the round. idk tho

**Comment by SgtxRootbeer** [comment]
@MrDoomBringer is right, the gameplay might be nice if we kept cloning records for suicide in

**Comment by Iamgoofball** [comment]
@MrDoomBringer If you want to get back into the round, go out an airlock. The Suicide verb is specifically a "do not revive" method of exiting the round.

**Comment by JJRcop** [comment]
You can also punch yourself into crit

**Comment by SgtxRootbeer** [comment]
@Iamgoofball alternatively, ghost, or suicide a second time, or dont get pre-scanned, or delete your scan?

**Comment by Kromgar** [comment]
I can imagine there might be situations where you are about to be converted by cult but you manage to break free trapped with nowhere else to go. You are scanned in autocloner and you dont have the time to punch yourself to death. So you commit suicide.

**Comment by george99g** [comment]
There's a compromise of stopping autoprocess from cloning you after a suicide, but still letting a geneticist manually hit the button.

**Comment by Tlaltecuhtli** [comment]
@kromgar that sounds like emergent gameplay

**Comment by AlexTheSysop42** [comment]
@Kromgar Plot twist: they just shade you anyway

or can you not shade suicides?

**Comment by MrDoomBringer** [comment]
@Iamgoofball what if there's not an airlock available? Also where does the game say that you shouldn't be revivable? You can get cloned and defibbed as easily as anyone else

**Comment by uraniummeltdown** [comment]
Not only is this a terrible change, it's unrealistic for the cloner to magically detect you suicided and remove your scan

**Comment by AlexTheSysop42** [comment]
>Also where does the game say that you shouldn't be revivable?

I think the confirmation window says it, could be wrong though

**Comment by george99g** [comment]
> Not only is this a terrible change, it's unrealistic for the cloner to magically detect you suicided and remove your scan

It doesn't actually remove the scan. It just says your brain data is malformed and doesn't let you clone the record.
The record still exists.

**Comment by george99g** [comment]
> I think the confirmation window says it, could be wrong though

You're thinking of the ghost window while you're alive, I think:
screenshot at 2018-03-27 15-55-24

**Comment by AlexTheSysop42** [comment]
`var/confirm = alert("Are you sure you want to commit suicide?", "Confirm Suicide", "Yes", "No")`

ok yeah looks like there isn't a warning

**Comment by uraniummeltdown** [comment]
When I suicide, that means I want to die, I want to kill myself. It doesn't mean I want to exit the round, it doesn't mean I want to become unrevivable.
@george99g that's still stupid, your brain data was fine when you scanned yourself but it magically becomes malformed when the cloner magically detects you suicided? unrealistic and unnecessary

**Comment by vuonojenmustaturska** [comment]
@Kromgar if you're suiciding/disconnecting to avoid being converted, just ask an admin for a conversion antag permaban, that will give your body to ghosts and won't screw over the cult side.

**Comment by Iatots** [comment]
Add a seppuku verb to kill yourself specifically because the situation demands it and not because you don't want to play anymore.

**Comment by george99g** [comment]
> It doesn't mean I want to exit the round, it doesn't mean I want to become unrevivable.

The rest of the game assumes you want to become unrevivable. The scanner will refuse to make a scan and the defib will refuse to revive you.
This seems more like a bug that's making the game inconsistent than an intended feature.

**Comment by AlexTheSysop42** [comment]
@vuonojenmustaturska he's talking about suiciding after escaping, which should be allowed to screw over the cult side because they let you escape (and it's very hard to free yourself, nigh impossible if they're careful)

Also, you can't suicide if you're cuffed

**Comment by GrayRachnid** [comment]
I'd like to remind everyone that this pr only exists because there's an issue and it's marked as a bug

**Comment by iskyp** [comment]
add a dnr verb like goon has, so you can set whether or not you want to stay dead

**Comment by 81Denton** [comment]
#36212 was sadly closed, it would make this a bit easier.

**Comment by Kromgar** [comment]
@Bawhoppen I thought the autoscan cloner implanted wireless chips into your brain

**Comment by george99g** [comment]
> add a dnr verb like goon has, so you can set whether or not you want to stay dead

That would have to be another PR. This one's just a bugfix.

File: code/datums/mind.dm [hunk 1]
```diff
@@ -774,6 +774,30 @@
...
+		//Do the same for diskettes
+		for(var/obj/item/disk/data/D in world)
+			if(D.fields["ckey"] && D.fields["ckey"] == user.ckey)
+				D.fields["suiciding"] = TRUE
+
+		return
+
+	log_game("[key_name(user)] committed suicide at [get_area(user)][COORD(T)] as [user.type].")
+
+	return
```
Comment (ninjanomnom) [inline code/datums/mind.dm]: this return is unnecesary

**Review by ninjanomnom — DISMISSED** [review]

**Comment by ExcessiveUseOfCobblestone** [comment]
This whole PR is unnecessary, I've explained in a previous PR that I think @XDTM did that the cloner was actually reverting your suicide status. Simply removing that line would fix the issue.

**Comment by ExcessiveUseOfCobblestone** [comment]
See https://github.com/tgstation/tgstation/pull/36212#pullrequestreview-102410045

**Comment by ExcessiveUseOfCobblestone** [comment]
Also breaking rules shouldn't be a reason to avoid fixing an inconsistency with the verb ...

As it is, suicide is a "I do not want to return to the round as that character" verb in every instance but this edge case. Bikeshedders btfo or change it in a separate pr to be consistent.

**Comment by george99g** [comment]
> This whole PR is unnecessary for the issue it's trying to fix, I've explained in a previous PR that I think @XDTM did that the cloner was actually reverting your suicide status. Simply removing that line would fix the issue.

IIRC @GrayRachnid tried that change and it didn't quite work, but I don't remember the details as to why.

Besides, that fix would eject the occupant if they're a suicide, while this will just stop them from getting into the pod in the first place, in addition to giving the geneticists some notification about what's happening.

**Comment by ExcessiveUseOfCobblestone** [comment]
@george99g it should already be throwing an error, the issue is explicitly being scanned while ALIVE **then** suiciding.

**Comment by george99g** [comment]
> @george99g it should already be throwing an error, the issue is explicitly being scanned while ALIVE then suiciding.

This fixes exactly that issue, though.

**Comment by ExcessiveUseOfCobblestone** [comment]
@george99g which goes back to my initial comment...

Make sure removing that line DOESN'T work

@Iamgoofball since it's a registered bug, if I see anyone using it to lol nope deconversion they'll get conversion mode banned under exploiting.

**Comment by ExcessiveUseOfCobblestone** [comment]
use runtime station.

ill see if there's an easier way because I don't like the whole "anytime you suicide it magically syncs your records" thing.

**Comment by ExcessiveUseOfCobblestone** [comment]
It doesn't need to, once they pop out they'll get the reason if they try to rescan.

I can't forsee any edge-cases unless people have been under the impression that the ghost proc already does such and have coded checks for it, and at that point you're really just fixing a plethora of bugs we didn't even know were bugs.

**Comment by ExcessiveUseOfCobblestone** [comment]
Regardless /how/ it's done though, I think the whole auto-updating records thing is lame and should  be removed (that's a feature sliding in with a bugfix)

**Comment by george99g** [comment]
> that's a feature sliding in with a bugfix

I don't know how this counts as a feature, unless you want me to remove any indication from the UI about why it's not letting you clone the person.

**Comment by XDTM** [comment]
I'm not sure where cobby got the impression that records would auto-update in this PR

File: code/datums/mind.dm [hunk 2]
```diff
@@ -774,6 +774,28 @@
...
+
+	if(ishuman(user))
+		var/mob/living/carbon/human/H = user
+		log_game("[key_name(user)] (job: [H.job ? "[H.job]" : "None"]) committed suicide at [get_area(user)][COORD(T)].")
+
+		//Go through all cloning computers and mark their records as suicided
+		for(var/obj/machinery/computer/cloning/C in world)
+			var/datum/data/record/R = find_record("ckey", user.ckey, C.records)
+			if(R)
+				R.fields["suiciding"] = TRUE
```
Comment (ExcessiveUseOfCobblestone) [inline code/datums/mind.dm]: @XDTM this is what I was seeing, just not reading very much into it

**Comment by ExcessiveUseOfCobblestone** [comment]
My bad, it's explicitly within the cloner and disks.

However, looping through literally every cloning computer and data disk for every suicide is crazy.

**Simply store that someone has suicided on the ghost and have it check said var on cloning**

**Comment by george99g** [comment]
> However, looping through literally every cloning computer and data disk for every suicide is crazy.

I don't think it's _that_ crazy, considering suicides, unless the player decides to use lavaland roles to suicide with them, would only happen at most once per player. I can agree with you about the data disks, though. Someone could quite easily make tons of them. I should figure out a way to avoid that.

> Simply store that someone has suicided on the ghost and have it check said var on cloning

I'll see what I can do with a var on ghosts and open an alternative PR.

**Comment by XDTM** [comment]
the ghost verb decouples the ghost from the body - might be as easy as copying that

**Comment by george99g** [comment]
> I can agree with you about the data disks, though. Someone could quite easily make tons of them. I should figure out a way to avoid that.

I was thinking of adding a list of references to disks in the records or using the disks as a doubly-linked list, but that's going to be a nightmare to debug and would probably require refactoring the data disks to not turn the code into a pile of spaghetti.

Quite frankly, I can't be bothered.

**Pull request merged**